### PR TITLE
nixos/prometheus-dnssec-exporter: init

### PR DESCRIPTION
**Pull request by SebastianWendel**
Prometheus DNSSEC Exporter package, module and test

## Description of changes

This PR introduces the `prometheus-dnssec-exporter` package and its accompanying NixOS module `services.prometheus.exporters.dnssec` to the Nixpkgs collection. The [prometheus-dnssec-exporter](https://github.com/chrj/prometheus-dnssec-exporter) is a tool designed to check the validity and expiration of DNSSEC signatures and expose these metrics for Prometheus, making it a valuable asset for monitoring the health and security of DNS infrastructure.

## Things done

- Added `prometheus-dnssec-exporter` package, allowing users to easily deploy this tool in a Nix environment.
- Created a NixOS module `services.prometheus.exporters.dnssec`, facilitating simple configuration and deployment of the service within NixOS systems.
- Implemented a NixOS test for the `services.prometheus.exporters.dnssec` module to ensure its correct operation and integration within the NixOS ecosystem.

---

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SebastianWendel** [comment]
@SuperSandro2000 since you have a dns focus, could you take a look at my PR and give feedback if necessary? i would be pleased if my contribution is adopted.

File: pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    rev = "b638685ed8d5919a88b45e85b3aec702f0fcc393";
+    hash = "sha256-SGoQKSgTRfSyA65xEZ9P7Z956sLMhB88h3HaXmFywiQ=";
+  };
+
+  vendorHash = "sha256-u7X8v7h1aL8B1el4jFzGRKHvnaK+Rz0OCitaC6xgyjw=";
+
+  doCheck = false;
+
+  meta = with lib; {
+    inherit (src.meta) homepage;
```
Comment (SuperSandro2000) [inline pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix:18]:
```suggestion
    homepage = "https://github.com/chrj/prometheus-dnssec-exporter";
```
that is easier to click on in review

File: pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  src = fetchFromGitHub {
+    owner = "chrj";
+    repo = "prometheus-dnssec-exporter";
+    rev = "b638685ed8d5919a88b45e85b3aec702f0fcc393";
+    hash = "sha256-SGoQKSgTRfSyA65xEZ9P7Z956sLMhB88h3HaXmFywiQ=";
+  };
+
+  vendorHash = "sha256-u7X8v7h1aL8B1el4jFzGRKHvnaK+Rz0OCitaC6xgyjw=";
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix:15]: Running https://github.com/chrj/prometheus-dnssec-exporter/blob/master/main_test.go does not work?

Comment (SebastianWendel) [reply pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix:15]: Good point. I must confess that I had probably mistakenly taken that from another package. I'll check it out.

Comment (SebastianWendel) [reply pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix]: Of course they worked as expected!

**Comment by SebastianWendel** [comment]
@SuperSandro2000 I hope everything should now be implemented correctly. Thanks for the support!

I still have a question about the version attribute, @eclairevoyant had suggested this to me as a change. The date and unstable are understandable, but what can I interpret the zero as? I looked in the manual but didn't find anything suitable or searched incorrectly. Do you have a source where I can read more about versioning? Or is this simply the version tag of the nix package definition?

**Comment by eclairevoyant** [comment]
> what can I interpret the zero as?

The versioning comes from [the contribution guidelines](https://github.com/NixOS/nixpkgs/tree/master/pkgs#package-naming):
> Example: Given a project had its latest releases `2.2` in November 2021, and `3.0` in January 2022, a commit authored on March 15, 2022 for an upcoming bugfix release `2.2.1` would have `version = "2.2-unstable-2022-03-15"`.
> 
> - If a project has no suitable preceding releases - e.g., no versions at all, or an incompatible versioning / tagging schema - then the latest upstream version in the above schema should be `0`.

File: pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  src = fetchFromGitHub {
+    owner = "chrj";
+    repo = "prometheus-dnssec-exporter";
+    rev = "b638685ed8d5919a88b45e85b3aec702f0fcc393";
+    hash = "sha256-SGoQKSgTRfSyA65xEZ9P7Z956sLMhB88h3HaXmFywiQ=";
+  };
+
+  vendorHash = "sha256-u7X8v7h1aL8B1el4jFzGRKHvnaK+Rz0OCitaC6xgyjw=";
+
```
Comment (cryptix) [inline pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix:14]:
```suggestion
  enableParallelBuilding = true;
```

**Review by cryptix — APPROVED** [review]
I ran the tests and they did what's on the tin. Looks great, as far as I can tell.

Only have a small thing about parallel building, might be small in grand scheme of things but every bit helps, i think. Sandro already picked out the raisins.

File: pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/by-name/pr/prometheus-dnssec-exporter/package.nix:14]: That's the default https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/go/module.nix#L34

**Pull request merged**